### ISSUE TITLE
VRAM Requirement on Training GRACE

### ISSUE THREAD

**Issue by junsiknss** [Thartvigsen/GRACE#7]
Hello. I created an issue because I have a question about the GRACE paper.

In the paper you said 'GRACE only consists of ~2% of the pretrained model's total parameters.'

So if the original model is running for inference, and loading the ~2% additional parameters for GRACE on GPU, can GRACE proceed with training GRACE Values on the inputs only with this resources? 

Or does GRACE need more memory to learn about new key-valu pairs, just as LLM requires much more VRAM to train and finetune than to inference?

It would be very helpful if there were any specific GPU VRAM requirements identified during the training GRACE

Thanks.

**Comment by Thartvigsen** [comment]
GRACE requires backprop through the model, so it does require more VRAM to train than for inference! In principle, it can be implemented for cheaper VRAM costs because we don't need gradients for the whole model, but for the current implementation it makes sense to assume you need the same VRAM for GRACE as you would for finetuning because learning the values is done via a loss computed on the model's outputs.

**Comment by junsiknss** [comment]
Thanks for the answer!
If I were to modify GRACE's implementation to not load all the gradients and parameters, which part of GRACE's components would I need to modify?

**Comment by Thartvigsen** [comment]
Honestly, I'm not 100% certain for implementation. The key idea would be that we only need gradients from the last layer to the GRACE values being learned. If we then edit a later layer, we don't need gradients for any layers before the edited layer. That said, does pytorch already handle this case? If not automatic, maybe it's helpful to set ```requires_grad=False``` to all parameters in all layers lower than where GRACE is added? There's no gradients to compute for those anyways, so maybe that's already handled.

**Comment by junsiknss** [comment]
That was helpful, thank you.
I'll close the issue
